src/server/pps/server/api_server.go

```diff
@@ -772,6 +772,31 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *ppsclient.Creat
 	for _, input := range request.Inputs {
 		provenance = append(provenance, input.Repo)
 	}
+	if !request.Update { // repo exists if it's an update
+		// This function needs to return newErrPipelineExists if the pipeline
+		// already exists
+		if _, err := a.InspectPipeline(
+			ctx,
+			&ppsclient.InspectPipelineRequest{Pipeline: request.Pipeline},
+		); err == nil {
+			return nil, newErrPipelineExists(request.Pipeline.Name)
+		}
+		if _, err := pfsAPIClient.CreateRepo(
+			ctx,
+			&pfsclient.CreateRepoRequest{
+				Repo:       repo,
+				Provenance: provenance,
+			}); err != nil {
+			return nil, err
+		}
+		defer func() {
+			if retErr != nil {
+				// we ignore the error here because the function has already
+				// errored, if this fails there's nothing more we can do
+				pfsAPIClient.DeleteRepo(ctx, &pfsclient.DeleteRepoRequest{Repo: repo})
```

Comment:
If we can't recover / the state is undefined, it might be useful to report this to the user.